### PR TITLE
[FLINK-8761] [quickstarts] Big improvements to the quickstart experience

### PR DESCRIPTION
**Pull request by StephanEwen**
## What is the purpose of the change

Various improvements to the quickstarts, each in one commit.

The main improvement is that the quickstarts now work out of the box for both running in-IDE and packaging JARs properly. Without need to add something like `-Pbuild-jar` on the command line, which many users seem to get confused about or not understand at all (given the number of users that have big dependencies in their application Jar, including Flink core, Scala, Hadoop, etc)

## Changelog

  - [FLINK-8764] [quickstarts] Make quickstarts work out of the box for IDE and JAR packaging
  - [FLINK-8762] [quickstarts] Make 'StreamingJob' the default main class and remove WordCount example from the quickstart.
  - [FLINK-8763] [quickstarts] Remove obsolete Dummy.java classes from quickstart projects.
  - [FLINK-8765] [quickstarts] Simplify quickstart properties, making the quickstarts a bit simpler.
  - [FLINK-8766] [quickstarts] Pin scala runtime version for Java Quickstart
  - [FLINK-8767] [quickstarts] Set the maven.compiler.source and .target properties for Java Quickstart

  - [hotfix] [quickstarts] Fix header and package declaration order.
  - [hotfix] [quickstarts] Fix block comments in program stubs.

## Verifying this change

  - Building and installing the quickstarts
  - Setting up a project from the 1.5-SNAPSHOT quickstarts
  - Run sample in IDE (should work out of the box)
  - Package a jar, check what gets included

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no)**
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no)**
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no)**
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-quickstart/flink-quickstart-java/src/main/resources/archetype-resources/pom.xml [hunk 1]
```diff
@@ -50,181 +51,113 @@ under the License.
...
 			<groupId>org.apache.flink</groupId>
 			<artifactId>flink-java</artifactId>
 			<version>${flink.version}</version>
+			<scope>provided</scope>
 		</dependency>
 		<dependency>
-			<!-- This dependency is required to actually execute jobs. It is currently pulled in by
-				flink-streaming-java, but we explicitly depend on it to safeguard against future changes. -->
 			<groupId>org.apache.flink</groupId>
-			<artifactId>flink-clients_${scala.binary.version}</artifactId>
```
Comment (zentol) [inline flink-quickstart/flink-quickstart-java/src/main/resources/archetype-resources/pom.xml]: is this just for simplification or do we no longer expect that flink-streaming-java looses it's dependency on flink-clients?

**Comment by zentol** [comment]
The [quickstart docs|https://ci.apache.org/projects/flink/flink-docs-master/quickstart/java_api_quickstart.html] must be updated as they still list now deleted classes.

File: flink-quickstart/flink-quickstart-java/src/main/resources/archetype-resources/pom.xml [hunk 2]
```diff
@@ -50,181 +51,113 @@ under the License.
...
 
-		<!-- If you want to use Java 8 Lambda Expressions uncomment the following lines -->
-		<!--
 		<pluginManagement>
 			<plugins>
+
+				<!-- If you want to use Java 8 Lambda Expressions uncomment the following lines -->
+				<!--
 				<plugin>
 					<artifactId>maven-compiler-plugin</artifactId>
```
Comment (zentol) [inline flink-quickstart/flink-quickstart-java/src/main/resources/archetype-resources/pom.xml]: update the source/target configuration parameter to use `java.version`?

File: flink-quickstart/flink-quickstart-java/src/main/resources/archetype-resources/pom.xml [hunk 3]
```diff
@@ -50,181 +51,113 @@ under the License.
...
 		</dependency>
+
+		<!-- Add connector dependencies here. They must be in the default scope (compile). -->
+
+		<!-- Example:
+
 		<dependency>
 			<groupId>org.apache.flink</groupId>
-			<artifactId>flink-streaming-java_${scala.binary.version}</artifactId>
+			<artifactId>flink-connector-kafka-0.10_${scala.binary.version}</artifactId>
```
Comment (zentol) [inline flink-quickstart/flink-quickstart-java/src/main/resources/archetype-resources/pom.xml]: we may want to use a different example connector as the compilation fails with this one due to a dependency convergence error.

**Comment by zentol** [comment]
Building the quickstart project with the `Execute maven goal` in IntelliJ `Maven Projects` window causes `flink-dist` to be included in the jar. This also happened when disabling the IntelliJ profile.

When going through the maven terminal in IntelliJ the jar is properly built.

I don't know whether this is really a problem for us.

File: flink-quickstart/flink-quickstart-java/src/main/resources/archetype-resources/pom.xml (hunk 3 above)
Comment (StephanEwen) [reply flink-quickstart/flink-quickstart-java/src/main/resources/archetype-resources/pom.xml]: Interesting - in my setup, dependency convergence is not activated by default. Did you explicitly activate it?

File: flink-quickstart/flink-quickstart-java/src/main/resources/archetype-resources/pom.xml (hunk 2 above)
Comment (StephanEwen) [reply flink-quickstart/flink-quickstart-java/src/main/resources/archetype-resources/pom.xml]: good point, will do

**Comment by StephanEwen** [comment]
Thanks for checking this out!

Concerning the removed `flink-clients` dependency - that was done on purpose. Here is my motivation for that:
  - For the 'provided' API dependency, it should actually not be there.
  - It mainly matters to the in-Eclipse execution, which needs a 'provided' runtime.
  - I would prefer to remove it now, because the dependencies probably change anyways in the future.

The proposal I want to make for the next Flink version is to have something like `flink-all` (or `flink-base`) which exactly refers to what is in the `flink-dist` jar (and `flink-dist` draws its jar from there). We set that as a provided dependency anywhere (connector, library, quickstart), we have exactly everything provided that will available through Flink's runtime, and nothing that will not be there. Easiest way to keep these things in sync.

Now I am unsure if `flink-dist` can currently take a `flink-all` like role, because it declares a lot of additional provided dependencies, for the sake of putting them into `opt` or `examples`. Now, according to the maven dependency management, a transitive provided dependency of a provided dependency is not propagated (in which case it would be okay to use flink-dist), but I am not sure we want to rely that across maven command line, IntelliJ and Eclipse, etc.

The last sentence incidentally just made me realize that I should probably change the `flink-dist` dependency in the IDEA profile to `flink-streaming-java` / `flink-clients` for exactly that reason...

File: flink-quickstart/flink-quickstart-java/src/main/resources/archetype-resources/pom.xml (hunk 3 above)
Comment (zentol) [reply flink-quickstart/flink-quickstart-java/src/main/resources/archetype-resources/pom.xml]: no i didn't activate it. I only un-commented the dependency and tried to build it.

It's a good question why this happened; after all the quickstart projects do not inherit from Flink.

Comment (zentol) [reply flink-quickstart/flink-quickstart-java/src/main/resources/archetype-resources/pom.xml]: hmm. I created another quickstart to reproduce it but it worked this time.

**Comment by StephanEwen** [comment]
Thanks for checking it out and the feedback.

Merging this...

  > Building the quickstart project with the Execute maven goal method in IntelliJs Maven Projects window causes flink-dist to be included in the jar. This also happened when disabling the IntelliJ profile.
I think that is not a big issue...

**Comment by StephanEwen** [comment]
Merged in 647c552a26cbe5f37dfb1d69f26574ef0853fba3

**Pull request closed without merging**